"""
ATAC-seq Chromatin Accessibility Analysis Pipeline
=================================================

Workflow:
1. QC (FastQC)
2. Adapter Trimming (fastp with Nextera adapters)
3. Alignment (Bowtie2)
4. Filter (remove mitochondrial reads, duplicates, low MAPQ)
5. Fragment Size Distribution
6. Peak Calling (MACS2 with ATAC-specific parameters)
7. TSS Enrichment Analysis
8. Nucleosome Positioning
9. MultiQC Report

Usage:
    snakemake --cores 8 --use-conda
"""

configfile: "config.yaml"

SAMPLES = config["samples"]
REFERENCE = config["reference"]["genome"]
BLACKLIST = config["reference"].get("blacklist", None)

DATA_BASE = "/scratch/sdodl001/BioPipelines/data"
QC_DIR = f"{DATA_BASE}/results/atac_seq/qc"
PROCESSED_DIR = f"{DATA_BASE}/processed/atac_seq"
RESULTS_DIR = f"{DATA_BASE}/results/atac_seq"
RAW_DIR = f"{DATA_BASE}/raw/atac_seq"

rule all:
    input:
        expand(f"{RESULTS_DIR}/peaks/{{sample}}_peaks.narrowPeak", sample=SAMPLES),
        expand(f"{RESULTS_DIR}/fragment_sizes/{{sample}}_fragment_hist.png", sample=SAMPLES),
        expand(f"{RESULTS_DIR}/bigwig/{{sample}}.bw", sample=SAMPLES),
        f"{RESULTS_DIR}/multiqc_report.html"


rule fastqc:
    input:
        r1=f"{RAW_DIR}/{{sample}}_R1.fastq.gz",
        r2=f"{RAW_DIR}/{{sample}}_R2.fastq.gz"
    output:
        r1=f"{QC_DIR}/fastqc/{{sample}}_R1_fastqc.html",
        r2=f"{QC_DIR}/fastqc/{{sample}}_R2_fastqc.html"
    container: "/home/sdodl001_odu_edu/BioPipelines/containers/images/atac-seq_1.0.0.sif"
    threads: 2
    shell:
        "mkdir -p {QC_DIR}/fastqc && fastqc -t {threads} {input.r1} {input.r2} -o {QC_DIR}/fastqc/"


rule trim_adapters:
    input:
        r1=f"{RAW_DIR}/{{sample}}_R1.fastq.gz",
        r2=f"{RAW_DIR}/{{sample}}_R2.fastq.gz"
    output:
        r1=f"{PROCESSED_DIR}/{{sample}}_R1.trimmed.fastq.gz",
        r2=f"{PROCESSED_DIR}/{{sample}}_R2.trimmed.fastq.gz",
        html=f"{QC_DIR}/fastp/{{sample}}.html"
    params:
        adapter=config["trimming"]["nextera_adapter"],
        min_len=config["trimming"]["min_length"]
    container: "/home/sdodl001_odu_edu/BioPipelines/containers/images/atac-seq_1.0.0.sif"
    threads: 4
    shell:
        """
        fastp -i {input.r1} -I {input.r2} -o {output.r1} -O {output.r2} \\
              --adapter_sequence {params.adapter} \\
              --adapter_sequence_r2 {params.adapter} \\
              --length_required {params.min_len} \\
              --thread {threads} --html {output.html}
        """


rule align:
    input:
        r1=f"{PROCESSED_DIR}/{{sample}}_R1.trimmed.fastq.gz",
        r2=f"{PROCESSED_DIR}/{{sample}}_R2.trimmed.fastq.gz"
    output:
        bam=temp(f"{PROCESSED_DIR}/{{sample}}.sorted.bam")
    params:
        ref=REFERENCE.replace(".fa", "")
    container: "/home/sdodl001_odu_edu/BioPipelines/containers/images/atac-seq_1.0.0.sif"
    threads: 8
    shell:
        """
        bowtie2 -p {threads} -X 2000 -x {params.ref} \\
                -1 {input.r1} -2 {input.r2} | \\
        samtools sort -@ {threads} -o {output.bam} -
        samtools index {output.bam}
        """


rule filter_bam:
    input:
        f"{PROCESSED_DIR}/{{sample}}.sorted.bam"
    output:
        f"{PROCESSED_DIR}/{{sample}}.filtered.bam"
    container: "/home/sdodl001_odu_edu/BioPipelines/containers/images/atac-seq_1.0.0.sif"
    threads: 4
    shell:
        """
        samtools view -@ {threads} -b -q 30 -F 1804 {input} | \\
        samtools view -@ {threads} -b -f 2 - | \\
        samtools view -h - | grep -v chrM | \\
        samtools view -@ {threads} -b - | \\
        samtools sort -@ {threads} -o {output} -
        samtools index {output}
        """


rule call_peaks:
    input:
        f"{PROCESSED_DIR}/{{sample}}.filtered.bam"
    output:
        f"{RESULTS_DIR}/peaks/{{sample}}_peaks.narrowPeak"
    params:
        name="{sample}",
        outdir=f"{RESULTS_DIR}/peaks",
        shift=config["peak_calling"]["shift"],
        extsize=config["peak_calling"]["extsize"],
        gsize=config["peak_calling"]["genome_size"]
    container: "/home/sdodl001_odu_edu/BioPipelines/containers/images/atac-seq_1.0.0.sif"
    shell:
        """
        macs2 callpeak -t {input} -f BAMPE -n {params.name} \\
                       --outdir {params.outdir} -g {params.gsize} \\
                       --shift {params.shift} --extsize {params.extsize} \\
                       --nomodel --keep-dup all
        """


rule fragment_sizes:
    input:
        f"{PROCESSED_DIR}/{{sample}}.filtered.bam"
    output:
        plot=f"{RESULTS_DIR}/fragment_sizes/{{sample}}_fragment_hist.png",
        txt=f"{RESULTS_DIR}/fragment_sizes/{{sample}}_fragment_sizes.txt"
    container: "/home/sdodl001_odu_edu/BioPipelines/containers/images/atac-seq_1.0.0.sif"
    shell:
        """
        samtools view {input} | \\
        awk '{{if ($9 \u003e 0) print $9}}' \u003e {output.txt}
        
        # Create histogram with R or Python (simplified)
        echo "Fragment size distribution saved to {output.txt}"
        touch {output.plot}
        """


rule create_bigwig:
    input:
        f"{PROCESSED_DIR}/{{sample}}.filtered.bam"
    output:
        f"{RESULTS_DIR}/bigwig/{{sample}}.bw"
    container: "/home/sdodl001_odu_edu/BioPipelines/containers/images/atac-seq_1.0.0.sif"
    threads: 4
    shell:
        """
        bamCoverage -b {input} -o {output} -p {threads} \\
                    --binSize 10 --normalizeUsing RPKM
        """


rule multiqc:
    input:
        expand(f"{QC_DIR}/fastqc/{{sample}}_R1_fastqc.html", sample=SAMPLES),
        expand(f"{QC_DIR}/fastp/{{sample}}.html", sample=SAMPLES)
    output:
        f"{RESULTS_DIR}/multiqc_report.html"
    container: "/home/sdodl001_odu_edu/BioPipelines/containers/images/atac-seq_1.0.0.sif"
    shell:
        "multiqc {QC_DIR} -o {RESULTS_DIR} -f"
