"""
DNA-seq Variant Calling Pipeline
==================================

A comprehensive GATK best practices pipeline for germline variant calling.

Workflow steps:
1. Quality Control (FastQC)
2. Read Trimming (fastp)
3. Alignment (BWA-MEM)
4. Mark Duplicates (Picard)
5. Base Quality Score Recalibration (GATK)
6. Variant Calling (GATK HaplotypeCaller)
7. Variant Filtering
8. Annotation (SnpEff)
9. MultiQC Report

Usage:
    snakemake --cores 4 --use-conda
"""

import os
from pathlib import Path

# Load configuration
configfile: "config.yaml"

# Get samples from config
SAMPLES = config["samples"]
REFERENCE = config["reference"]["genome"]
KNOWN_SITES = config["reference"]["known_sites"]

# Output directories
DATA_BASE = "/scratch/sdodl001/BioPipelines/data"
QC_DIR = f"{DATA_BASE}/results/qc"
PROCESSED_DIR = f"{DATA_BASE}/processed"
RESULTS_DIR = f"{DATA_BASE}/results"
RAW_DIR = f"{DATA_BASE}/raw/dna_seq"
RAW_DIR = f"{DATA_BASE}/raw/dna_seq"

# Final outputs
rule all:
    input:
        # Final VCF files (skip annotation for now due to network issues)
        expand(f"{RESULTS_DIR}/vcf/{{sample}}.filtered.vcf.gz", sample=SAMPLES),
        expand(f"{RESULTS_DIR}/vcf/{{sample}}.filtered.vcf.gz.tbi", sample=SAMPLES),
        # MultiQC report
        f"{RESULTS_DIR}/multiqc_report.html"


# Quality Control - FastQC on raw reads
rule fastqc_raw:
    input:
        r1=f"{RAW_DIR}/{{sample}}_R1.fastq.gz",
        r2=f"{RAW_DIR}/{{sample}}_R2.fastq.gz"
    output:
        r1=f"{QC_DIR}/fastqc_raw/{{sample}}_R1_fastqc.html",
        r2=f"{QC_DIR}/fastqc_raw/{{sample}}_R2_fastqc.html"
    container: "/home/sdodl001_odu_edu/BioPipelines/containers/images/dna-seq_1.0.0.sif"
    threads: 2
    log:
        f"{QC_DIR}/logs/fastqc_raw/{{sample}}.log"
    shell:
        """
        fastqc -t {threads} {input.r1} {input.r2} -o {QC_DIR}/fastqc_raw/ 2> {log}
        """


# Trimming - fastp
rule trim_reads:
    input:
        r1=f"{RAW_DIR}/{{sample}}_R1.fastq.gz",
        r2=f"{RAW_DIR}/{{sample}}_R2.fastq.gz"
    output:
        r1=f"{PROCESSED_DIR}/{{sample}}_R1.trimmed.fastq.gz",
        r2=f"{PROCESSED_DIR}/{{sample}}_R2.trimmed.fastq.gz",
        html=f"{QC_DIR}/fastp/{{sample}}.html",
        json=f"{QC_DIR}/fastp/{{sample}}.json"
    container: "/home/sdodl001_odu_edu/BioPipelines/containers/images/dna-seq_1.0.0.sif"
    threads: 4
    log:
        f"{QC_DIR}/logs/trim/{{sample}}.log"
    shell:
        """
        fastp -i {input.r1} -I {input.r2} \
              -o {output.r1} -O {output.r2} \
              --thread {threads} \
              --html {output.html} \
              --json {output.json} \
              --detect_adapter_for_pe \
              --cut_right \
              --cut_right_window_size 4 \
              --cut_right_mean_quality 20 \
              --length_required 50 \
              2> {log}
        """


# Alignment - BWA-MEM
rule align_reads:
    input:
        r1=f"{PROCESSED_DIR}/{{sample}}_R1.trimmed.fastq.gz",
        r2=f"{PROCESSED_DIR}/{{sample}}_R2.trimmed.fastq.gz",
        ref=REFERENCE
    output:
        bam=temp(f"{PROCESSED_DIR}/{{sample}}.sorted.bam")
    params:
        rg="@RG\\tID:{sample}\\tSM:{sample}\\tPL:ILLUMINA\\tLB:{sample}\\tPU:unit1"
    container: "/home/sdodl001_odu_edu/BioPipelines/containers/images/dna-seq_1.0.0.sif"
    threads: 8
    log:
        f"{QC_DIR}/logs/align/{{sample}}.log"
    shell:
        """
        bwa mem -t {threads} -R '{params.rg}' {input.ref} {input.r1} {input.r2} 2> {log} | \
        samtools sort -@ {threads} -o {output.bam} - 2>> {log}
        samtools index {output.bam}
        """


# Mark Duplicates - GATK
rule mark_duplicates:
    input:
        bam=f"{PROCESSED_DIR}/{{sample}}.sorted.bam"
    output:
        bam=f"{PROCESSED_DIR}/{{sample}}.dedup.bam",
        metrics=f"{QC_DIR}/picard/{{sample}}.dedup_metrics.txt"
    container: "/home/sdodl001_odu_edu/BioPipelines/containers/images/dna-seq_1.0.0.sif"
    threads: 4
    log:
        f"{QC_DIR}/logs/dedup/{{sample}}.log"
    shell:
        """
        gatk MarkDuplicates \\
            -I {input.bam} \\
            -O {output.bam} \\
            -M {output.metrics} \\
            --CREATE_INDEX true \\
            --VALIDATION_STRINGENCY LENIENT \\
            2> {log}
        """


# Base Quality Score Recalibration - GATK
rule base_recalibration:
    input:
        bam=f"{PROCESSED_DIR}/{{sample}}.dedup.bam",
        ref=REFERENCE,
        known=KNOWN_SITES
    output:
        table=f"{PROCESSED_DIR}/{{sample}}.recal_data.table"
    container: "/home/sdodl001_odu_edu/BioPipelines/containers/images/dna-seq_1.0.0.sif"
    threads: 4
    log:
        f"{QC_DIR}/logs/bqsr/{{sample}}_recal.log"
    shell:
        """
        gatk BaseRecalibrator \
            -R {input.ref} \
            -I {input.bam} \
            --known-sites {input.known} \
            -O {output.table} \
            2> {log}
        """


rule apply_bqsr:
    input:
        bam=f"{PROCESSED_DIR}/{{sample}}.dedup.bam",
        ref=REFERENCE,
        table=f"{PROCESSED_DIR}/{{sample}}.recal_data.table"
    output:
        bam=f"{PROCESSED_DIR}/{{sample}}.recal.bam"
    container: "/home/sdodl001_odu_edu/BioPipelines/containers/images/dna-seq_1.0.0.sif"
    threads: 4
    log:
        f"{QC_DIR}/logs/bqsr/{{sample}}_apply.log"
    shell:
        """
        gatk ApplyBQSR \\
            -R {input.ref} \\
            -I {input.bam} \\
            --bqsr-recal-file {input.table} \\
            --create-output-bam-index true \\
            -O {output.bam} \\
            2> {log}
        """


# Variant Calling - GATK HaplotypeCaller
rule call_variants:
    input:
        bam=f"{PROCESSED_DIR}/{{sample}}.recal.bam",
        ref=REFERENCE
    output:
        vcf=f"{RESULTS_DIR}/vcf/{{sample}}.raw.vcf.gz",
        idx=f"{RESULTS_DIR}/vcf/{{sample}}.raw.vcf.gz.tbi"
    container: "/home/sdodl001_odu_edu/BioPipelines/containers/images/dna-seq_1.0.0.sif"
    threads: 4
    log:
        f"{QC_DIR}/logs/call_variants/{{sample}}.log"
    shell:
        """
        gatk HaplotypeCaller \
            -R {input.ref} \
            -I {input.bam} \
            -O {output.vcf} \
            --native-pair-hmm-threads {threads} \
            2> {log}
        """


# Variant Filtering
rule filter_variants:
    input:
        vcf=f"{RESULTS_DIR}/vcf/{{sample}}.raw.vcf.gz"
    output:
        vcf=f"{RESULTS_DIR}/vcf/{{sample}}.filtered.vcf.gz",
        idx=f"{RESULTS_DIR}/vcf/{{sample}}.filtered.vcf.gz.tbi"
    container: "/home/sdodl001_odu_edu/BioPipelines/containers/images/dna-seq_1.0.0.sif"
    log:
        f"{QC_DIR}/logs/filter/{{sample}}.log"
    shell:
        """
        gatk VariantFiltration \
            -V {input.vcf} \
            -O {output.vcf} \
            --filter-expression "QD < 2.0 || FS > 60.0 || MQ < 40.0 || MQRankSum < -12.5 || ReadPosRankSum < -8.0" \
            --filter-name "basic_snp_filter" \
            2> {log}
        """


# Annotation - SnpEff
rule annotate_variants:
    input:
        vcf=f"{RESULTS_DIR}/vcf/{{sample}}.filtered.vcf.gz"
    output:
        vcf=f"{RESULTS_DIR}/vcf/{{sample}}.annotated.vcf.gz",
        idx=f"{RESULTS_DIR}/vcf/{{sample}}.annotated.vcf.gz.tbi",
        html=f"{QC_DIR}/snpeff/{{sample}}.html"
    params:
        genome=config.get("snpeff_genome", "hg38")
    container: "/home/sdodl001_odu_edu/BioPipelines/containers/images/dna-seq_1.0.0.sif"
    log:
        f"{QC_DIR}/logs/annotate/{{sample}}.log"
    shell:
        """
        snpEff -v {params.genome} {input.vcf} \
            -stats {output.html} 2> {log} | \
        bgzip > {output.vcf}
        tabix -p vcf {output.vcf}
        """


# MultiQC - Aggregate all QC reports
rule multiqc:
    input:
        expand(f"{QC_DIR}/fastqc_raw/{{sample}}_R1_fastqc.html", sample=SAMPLES),
        expand(f"{QC_DIR}/fastp/{{sample}}.json", sample=SAMPLES),
        expand(f"{QC_DIR}/picard/{{sample}}.dedup_metrics.txt", sample=SAMPLES)
    output:
        f"{RESULTS_DIR}/multiqc_report.html"
    container: "/home/sdodl001_odu_edu/BioPipelines/containers/images/dna-seq_1.0.0.sif"
    log:
        f"{QC_DIR}/logs/multiqc.log"
    shell:
        """
        multiqc {QC_DIR} -o {RESULTS_DIR} -f 2> {log}
        """
